[#7] 로그인, 로그아웃 기능 구현

tilsong opened the pull request:
## PR 유형

- [ ] 버그 수정
- [x] 기능
- [ ] 코드 스타일
- [ ] 리팩토링
- [ ] 빌드
- [ ] CI/CD
- [ ] 문서
- [ ] 기타
  <br>

## 변경사항

### 변경 전
- 로그인, 로그아웃 기능 X
### 변경 후
- 로그인 및 로그아웃 기능이 추가되었습니다.
- API 요청 시 토큰이 사용될 수 있습니다. 이를 통해 토큰에 있는 인증 정보를 사용할 수 있습니다. 자세한 내용은 아래 사용 방법에 있습니다.

## 사용 방법

- JWT 토큰 
  - 설명
    - JWT 토큰은 API 요청 시 사용될 수 있는 인증 데이터이며, 액세스 토큰과 리프레시 토큰의 2종류가 있습니다. 
    - API 요청 시 헤더에 액세스 토큰을 추가하고, 서버의 필터는 이를 검증하여 요청을 허용합니다. 허용되지 않을 경우, 적절한 예외 메시지를 응답합니다.
    - 액세스 토큰 만료 시 리프레시 토큰을 통해 새로운 액세스 토큰을 발급 받을 수 있습니다. 
  - 액세스 토큰
    -  API 요청 시 사용되며,  "Authorization:  Bearer <토큰 내용>" 형식으로 HTTP 헤더에 추가되어야 합니다. 포스트맨을 사용할 경우, Headers에서 Key에 "Authorization", Value에 "Bearer <토큰 내용>" 형식으로 데이터를 넣으시면 됩니다.
    - 기본 유효 기간은 1일 이므로, 날짜가 지난 토큰은 사용할 수 없습니다. 필요하다면 JwtAuthenticationManager 파일의 generateAccessToken 메서드에서 날짜 수를 수정해서 늘릴 수 있습니다. jwt 설정 자체를 꺼둘 수도 있지만, 만약 코드에서 토큰 데이터(ex 토큰 내부에 담긴 userId)를 사용하신다면 날짜를 수정해서 사용하는 것을 추천 드립니다.
    - 토큰에 담긴 userId를 사용할 수 있습니다. UserController의 test 메서드를 참고하시면 됩니다. 토큰에 담고 싶은 다른 데이터가 있으시다면 말씀해주세요.
  - 리프레시 토큰
    - 액세스 토큰 재발급 시 사용됩니다. 유효 기간은 30일로 설정했습니다.

## 기타
- 아래의 의존성들이 추가되었습니다. 커밋 내용을 통해 의존성 추가 이유를 확인하실 수 있습니다.
  - io.jsonwebtoken:jjwt-api:0.11.5
  - io.jsonwebtoken:jjwt-impl:0.11.5
  - io.jsonwebtoken:jjwt-jackson:0.11.5
  - org.springframework.boot:spring-boot-starter-security

tilsong committed:
[#7] 로그인 기능 구현 - Controller, Service 로그인 데이터 검증 로직 추가

- 로그인 요청 dto 생성
  - validation 적용
- UserController에 로그인 메서드 추가
  - 응답 로직은 jwt 로직 완료 후 수정할 것.
- 로그인을 위한 LoginService 추가
  - id, password를 통한 유저 검증 로직 추가.
  - jwt 로직은 추가되지 않음.
- 유저 검증 시 발생할 수 있는 예외(NoSuchUserException, AuthenticationFailedException) 및 핸들러 메소드 추가

tilsong committed:
[#7] 로그인 기능 구현 - Controller, Service, AuthenticationManager

- 인증 토큰 생성을 위한 jwt 의존성 추가
  - io.jsonwebtoken:jjwt-api:0.11.5
  - io.jsonwebtoken:jjwt-impl:0.11.5
  - io.jsonwebtoken:jjwt-jackson:0.11.5
- 인증 관리를 위한 AuthenticationManeger 인터페이스와 jwt를 사용한 구현 클래스 생성 및 적용
- application.properties에 jwt의 비밀 키 추가

tilsong committed:
[#7] 로그인 기능 구현 - Filter

- Fileter를 통해 api 요청 시 jwt 토큰 인증 로직을 거치도록 구현
  - 특정 경로(/user/logIn, /user/signUp)은 인증 로직을 적용하지 않도록 로직 작성
  - 필터 내 예외에 대한 추가 처리 필요

tilsong committed:
[#7] 로그인 기능 구현 - Spring security

- 스프링 시큐리티 사용을 위한 'org.springframework.boot:spring-boot-starter-security' 의존성 추가

- AuthenticationManager 인터페이스 제거

- 스프링 시큐리티 설정
  - 회원 가입과 로그인 api 에는 인증 관련 로직이 적용되지 않도록 설정
  - 토큰에서 유저 id를 꺼내어 사용할 수 있도록 JwtAuthenticationFilter에 로직 추가
    - 컨트롤러의 /user/test 메서드를 통해 테스트 가능. 다음 커밋 때 지울 예정

tilsong committed:
[#7] 로그인 기능 구현 - Spring security

- JwtAuthenticationManager 에서 필요 없는 메서드 제거

- 필터에서 토큰 검증 시 발생하는 예외 처리 및 클라이언트로의 응답 설정
  - JwtAuthenticationEntryPoint  클래스 추가
  - 필터 및  필터 체인 로직 변경

tilsong committed:
[#7] 로그인 기능 구현 - Spring security

- 응답 데이터 인코딩 예외 수정

tilsong committed:
[#7] 로그인 기능 구현 - Spring security, Controller, Service

- 필터에서 발생하는 예외 처리 구조 변경
  - 기존 JwtAuthenticationEntryPoint에서 일괄 처리했던 필터에서의 예외를 필터 내부에서 핸들링하여 응답 처리할 수 있도록 수정

- 리프레시 토큰 발급 로직 추가
  - 토큰 엔티티 추가
  - 토큰 재발급 로직 및 로그아웃 처리를 위해 리프레시 토큰을 Repository에 저장
  - UserController에 토큰 재발급을 위한 메서드 추가

- 기존 LogInService를AuthenticationService로 변경
  - 로그인과 토큰 재발급 기능 수행
  - 리프레시 토큰과 관련된 로직 추가
  - 전역 핸들러에 Jwt 처리 관련해서 발생할 수 있는 예외 핸들러 추가

- JwtAuthenticationManager에 유틸 메서드 추가

tilsong committed:
[#7] 로그아웃 기능 구현 - Spring security, Controller, Service

tilsong committed:
Merge branch 'main' into feature/7

- 기존 PR을 반영한 main 브랜치의 내용을 merge하고, 이를 다시 커밋.

tilsong committed:
[#7] 로그인, 로그아웃 기능 구현 - 액세스 토큰 유효기간 수정

tilsong committed:
[#7] 로그인, 로그아웃 기능 구현 - 오타 수정 및 불필요한 로직 제거

f-lab-fi commented on src/main/java/com/directors/application/user/AuthenticationService.java:
```diff
@@ -0,0 +1,106 @@
+
+        User user = userRepository.findUserById(userId);
+
+        if (user == null) {
+            throw new NoSuchUserException(userId);
+        }
+
+        if (!pm.checkPassword(password, user.getPassword())) {
+            throw new AuthenticationFailedException(user.getUserId());
+        }
```
서비스들 로그인 페이지에서 로그인을 할 때를 잘 관찰해보면, **id가 존재하는지 여부**와 **password가 불일치하는지 여부**를 나눠서 알려주지 않습니다. 그 이유는 무엇일까요?

현재 은석님이 구현하신 코드에선 구분되고 있죠.

tilsong replied:
위의 예외를 사용한다면, NoSuchUserException 예외를 발생시키지 않는 id를 찾은 뒤 이를 무작위의 password와 조합하여 로그인 시도를 할 수 있기 때문입니다. 
따라서 로그인 시에는 id나 password 각각의 데이터에 대한 예외를 발생시켜서는 안 되고, 로그인 실패 예외로만 반환할 수 있어야 할 것 같습니다. 반영하여 수정하겠습니다!

f-lab-fi commented on src/main/java/com/directors/application/user/AuthenticationService.java:
```diff
@@ -0,0 +1,106 @@
+        String accessToken = request.accessToken();
+        String refreshToken = request.refreshToken();
+
+        // 토큰 유효성 검사 1 - 액세스 토큰과 리프레시 토큰의 userId 비교
+        String userId = jm.getUserIdByToken(accessToken);
+        if (!userId.equals(jm.getUserIdByToken(refreshToken))) {
+            throw new JwtException("유효하지 않은 토큰입니다.");
+        }
+
+        // 토큰 유효성 검사 2 - 토큰의 소유 user가 존재하는지 여부 검증
+        User userById = userRepository.findUserById(userId);
+        if (userById == null) {
+            throw new JwtException("유효하지 않은 토큰입니다.");
+        }
+
+        // 토큰 유효성 검사 3 - 레포지토리를 통한 리프레시 토큰 유효성 검증
+        Token tokenByTokenString = tokenRepository.findTokenByTokenString(refreshToken);
+        if (tokenByTokenString == null) {
+            throw new JwtException("유효하지 않은 토큰입니다.");
+        }
+
+        // 토큰 유효성 검사 4 - 유효기간을 통한 리프레시 토큰 유효성 검증
+        long refreshExpirationDay = jm.getExpirationDayByToken(refreshToken);
+        if (refreshExpirationDay < 0) {
+            throw new JwtException("유효하지 않은 토큰입니다.");
+        }
+
+        // 액세스 토큰 발급
+        accessToken = jm.generateAccessToken(userId);
+
+        // 유효 기간이 일주일 보다 적게 남았을 경우 리프레시 토큰 재발급
+        if (refreshExpirationDay < 7) {
+            tokenRepository.deleteToken(refreshToken);
+
+            refreshToken = jm.generateRefreshToken(userId);
+            Date expiration = jm.getExpirationByToken(refreshToken);
+
+            tokenRepository.saveToken(new Token(refreshToken, userId, expiration));
+        }
```
이렇게 메서드 하나에 쭉 이어서 작성하면서 각 코드 블록에 주석을 다는 것 보다는 private 메서드들로 나눠서 해당 메서드에 적절한 이름을 부여하는게 더 가독성이 좋아질 것 같습니다~

tilsong replied:
넵! 반영하겠습니다.

f-lab-fi commented on src/main/java/com/directors/presentation/common/security/JwtAuthenticationEntryPoint.java:
```diff
@@ -0,0 +1,31 @@
+
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+
+/**
+ * API 요청에 대한 JwtAuthenticationFilter에서의 토큰 검증이 실패할 경우,
+ * 스프링 시큐리티 컨텍스트에서 사용할 Authentication를 획득하지 못하므로 아래의 commence 메서드를 통해 요청이 처리됩니다.
+ * 만약 유저의 ROLE(USER, ADMIN..)을 설정한다면 다른 처리가 필요할 수 있지만,
+ * 현재는 USER ROLE만 있으므로 스프링 시큐리티에서 Authentication를 소유하고 있는지 여부만 판별하여 유저 인증 절차가 처리되도록 했습니다.
+ */
+// TODO: 유지 여부 검토 필요
```
네 일단 너무 고민하지 마시고 나중에 바꿔보시죠~

tilsong replied:
넵! 다른 기능 부터 구현해보겠습니다~!

puallab commented on src/main/java/com/directors/presentation/user/response/LogInResponse.java:
다른 record와 코드 스타일이 다른데 맞춰주면 좋을거같습니다!

tilsong replied:
네 수정하겠습니다!

puallab reviewed (approved):
jwt 토큰을 제가 개인적으로 학습을 해봐야할거같네요 ㅎㅎ 고생하셨습니다!
확인했습니다!

tilsong committed:
[#7] 로그인 기능 구현 - 피드백 반영

- logIn 시 발생할 수 있는 예외를  AuthenticationFailedException 하나로 변경

- refreshAuthentication 메서드의 내용을 분리하여 private 메서드로 만들고, 적절한 이름 붙이기

- LogInResponse의 코드 스타일 변경

tilsong committed:
[#7] 로그인 기능 구현 - 예외 처리

- JwtException를 래핑하는 JwtExceptionWrapper와 핸들러 추가
  - JwtException의 경우 메시지가 명확하게 전달되지 않으므로 이를 래핑하는 클래스 추가

-  IllegalArgumentException 예외를 위한 핸들러 추가

tilsong committed:
[#7] 로그인 기능 구현 - 예외 처리 수정

- JwtExceptionWrapper 예외를 지우고, 단순히 핸들러에서 JwtException의 메시지를 넣어주도록 함.
